### PR TITLE
Add background subtraction child algorithm to workflow

### LINKED ISSUE #27825: Add background subtraction option to workflow algorithm
**Original reporter:** Max at ISIS

Add the `ReflectometryBackgroundSubtraction` algorithm into the reflectometry workflow algorithms, with the entry point being `ReflectometryISISLoadAndProcess`, so that it can subsequently be added to the GUI.

### PR DESCRIPTION
**Pull request by gemmaguest**
**Report to:** Max at ISIS

This PR adds `ReflectometryBackgroundSubtraction` as a child algorithm to `ReflectometryReductionOne` and its parent algorithms `ReflectometryReductionOneAuto` and `ReflectometryISISLoadAndProcess`. The main changes involve:
- Refactoring to separate out the steps in `makeDetectorWS` to allow detectors of interest to be extracted without being summed. This allows the background correction to be done just to the detectors of interest.
- Although existing tests are all C++ I've added a python test file because one of the background subtraction methods requires a python algorithm.
- There is a minor fix to `ReflectometryBackgroundSubtraction` to avoid setting the output name for a child algorithm. The output name is irrelevant for the child algorithm and this failing if no output name was specified for the parent algorithm.
Note that:
- The workflow algorithm has `BackgroundProcessingInstructions` instead of `ProcessingInstructions`
- `InputWorkspaceIndexType` is hidden because this is set by the parent algorithm.
- `PeakRange` is taken from the main `ProcessingInstructions` in the parent algorithm, where required
- `SumPeak` is hidden - this is always false because we do summing later in the reduction.

Note that:

- The workflow algorithm has `BackgroundProcessingInstructions` instead of `ProcessingInstructions`
- `InputWorkspaceIndexType` is hidden because this is set by the parent algorithm.
- `PeakRange` is taken from the main `ProcessingInstructions` in the parent algorithm, where required
- `SumPeak` is hidden - this is always false because we do summing later in the reduction.
- This is intentionally just a first cut and full validation will be done by scientists once it is released.

This PR also fixes some issues where existing python algorithms tests:
- A problem in `CMakeLists.txt` meant that existing tests weren't being run
- When enabled there were then failures in `ReflectometryBackgroundSubtractionTest`. These are minor so are also fixed in this PR.

Fixes #27825

**To test:**

Requires access to the ISIS archive. An example script is below - set the default instrument to `OFFSPEC` to run this script. This runs the reduction on a workspace with and without a fake added background and checks that the background subtraction option in the reduction algorithm removes it.

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

# Load a run and transmission
AnalysisDataService.clear()
LoadNexus(Filename='OFFSPEC44956', OutputWorkspace='TOF_44956')
LoadNexus(Filename='OFFSPEC44937', OutputWorkspace='TOF_44937')

# Create a copy of the run with an added flat background 
nmonitors=3
nspec=772
datay=[0]*3 + [3]*(nspec-nmonitors)
CreateWorkspace(OutputWorkspace='flat_background', DataX=[5,100000], DataY=datay, NSpec=nspec, UnitX='TOF', ParentWorkspace='TOF_44956')
Plus(LHSWorkspace='TOF_44956', RHSWorkspace='flat_background', OutputWorkspace='TOF_44956_bkg')

# Run the reduction on the original workspace
ReflectometryISISLoadAndProcess(InputRunList='TOF_44956',ProcessingInstructions='395-405',ThetaLogName='Theta',MomentumTransferStep=0.02,
  FirstTransmissionRunList='44937', PolarizationAnalysis=False, GroupTOFWorkspaces=False,
  OutputWorkspaceBinned='IvsQ_binned_44956', OutputWorkspace='IvsQ_44956', OutputWorkspaceWavelength='IvsLam_44956')

# Run the reduction on the workspace with added background and apply the correction  
ReflectometryISISLoadAndProcess(InputRunList='TOF_44956_bkg',ProcessingInstructions='395-405',ThetaLogName='Theta',MomentumTransferStep=0.02,
  FirstTransmissionRunList='44937', PolarizationAnalysis=False, GroupTOFWorkspaces=False,
  SubtractBackground=True, BackgroundProcessingInstructions='500-600',
  OutputWorkspaceBinned='IvsQ_binned_cor_44956', OutputWorkspace='IvsQ_cor_44956', OutputWorkspaceWavelength='IvsLam_cor_44956')

# Corrected workspace matches original to 0.001
CompareWorkspaces(Workspace1='IvsQ_binned_44956', Workspace2='IvsQ_binned_cor_44956', Tolerance=0.001)
```
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/test/ReflectometryBackgroundSubtractionTest.py [hunk 1]
```diff
@@ -123,29 +123,31 @@ def test_peak_range_changes_with_index_type(self):
...
         self._assert_run_algorithm_throws(args)
 
-    def test_AveragePixelFit_error_peakRange_outside_spectra(self):
-        args = {'InputWorkspace' : 'workspace_with_peak',
-                'ProcesssingInstructions' : '1-7',
-                'BackgroundCalculationMethod' : 'AveragePixelFit',
-                'PeakRange' : '3-9',
-                'OutputWorkspace': 'output'}
-        self._assert_run_algorithm_throws(args)
+    # TODO: This test fails in python 2. It can be re-added when we
```
Comment (ConorMFinn) [inline Framework/Algorithms/test/ReflectometryBackgroundSubtractionTest.py]: Has an issue or PR been made for this as a reminder?

**Review by cailafinn — CHANGES_REQUESTED** [review]
Code looks good. Just a comment about a `TODO`. However, when running the provided script it does not seem to be able to perform a `CalculateFlatBackground` fit and throws an error:

```
Error in execution of algorithm Fit:
StartX and EndX values do not capture a range within the workspace interval.
CalculateFlatBackground: StartX and EndX values do not capture a range within the workspace interval.
Error in execution of algorithm CalculateFlatBackground:
CalculateFlatBackground: error (see log)
Error in execution of algorithm CreateTransmissionWorkspace:
CalculateFlatBackground: error (see log)
Error in execution of algorithm ReflectometryReductionOne:
CalculateFlatBackground: error (see log)
Error in execution of algorithm ReflectometryReductionOneAuto:
CalculateFlatBackground: error (see log)
Error in execution of algorithm ReflectometryISISLoadAndProcess:
CalculateFlatBackground: error (see log)
  at line 121 in '/home/conor/repos/mantid-review/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
  caused by line 417 in '/home/conor/repos/mantid-review/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISLoadAndProcess.py'
```

**Comment by gemmaguest** [comment]
> when running the provided script it does not seem to be able to perform a `CalculateFlatBackground` fit and throws an error:

Ah, sorry, this is an unrelated error in the reduction algorithm and is because there is an assumption somewhere about the instrument. Set the default instrument to `OFFSPEC` and it should work.

File: Framework/Algorithms/test/ReflectometryBackgroundSubtractionTest.py (hunk 1 above)
Comment (gemmaguest) [reply Framework/Algorithms/test/ReflectometryBackgroundSubtractionTest.py]: :+1: There is now: #27995

**Review by cailafinn — APPROVED** [review]
Code looks good and script runs without errors. Issue made for commented test. Nice one :+1:

**Pull request merged**